libc-bottom-half/sources/posix.c

```diff
@@ -18,96 +19,115 @@ int open(const char *path, int oflag, ...) {
 
 // See the documentation in libc.h
 int __wasilibc_open_nomode(const char *path, int oflag) {
-    const char *relative_path;
-    int dirfd = __wasilibc_find_relpath(path, &relative_path);
+    const char *abs;
+    char *relative_path;
+    int dirfd = __wasilibc_find_relpath(path, &abs, &relative_path);
 
     // If we can't find a preopen for it, indicate that we lack capabilities.
     if (dirfd == -1) {
         errno = ENOTCAPABLE;
         return -1;
     }
 
-    return __wasilibc_openat_nomode(dirfd, relative_path, oflag);
+    int ret = __wasilibc_openat_nomode(dirfd, relative_path, oflag);
+    free(relative_path);
```

Comment:
It's unfortunate to add a `malloc`+`free` on every syscall. Obviously in the case of real host syscalls these will be relatively cheap, but in virtualized WASI environments, this could potentially be significant. Another potential concern is that we may in the future offer a WASI libc mode where `free` is a no-op and we just leak everything, to save the code size and runtime overhead of a real `malloc`. Such approaches are ok with leaking a bit of memory when the program is setting up, but less ok with leaking during operations like `open` which may be called many times.

That said, I also recognize that this allows for greater fidelity with POSIX and existing programs. I don't yet have a good suggestion for what to do here. It's possible we can address these concerns by making some things optional. I'll think about this more.

Reply:
I was thinking about this and at a fundamental level I think that malloc has to be supported at least sometimes, for example if you have a preopen of `/`, chdir to `/a`, and then open `b/c`, then you'll want to open `a/b/c` relative to `/`, but that's clearly no longer a subslice of the original path `b/c` since we needed to add `a/` onto it.

I could add an optimization where if a subslice of the original path can be returned then it is, but I'd need to add another return value flag for whether the relative path is owned or not and whether it needs to be deallocated.

Reply:
One traditional (and sometimes problematic) solution to this issue to have an static internal buffer here that gets realloc'd to be as large as is needs to be for each call.   This would be in TLS to allow in to work across threads.  The limitation of course is that successive calls to this function will clobber the output.  However, given that this is strictly a libc-internal function and not user facing, I don't see that as a big problem in this case.  This single buffer would then get a free'd during atexit.

Reply:
Ah good idea! I think though that would cause issues with `link` and `rename` which need to call `__wasilibc_find_relpath` twice and call `*at` functions with both path simultaneously.

Alternatively `__wasilibc_find_relpath` could be changed to not do the `malloc` itself but instead require a buffer get passed in, and it could return `ERANGE` when the buffer is too small. There could be a helper function in `posix.c` which would use one of perhaps two globals which cache a global buffer to be used. That would at least remove `malloc` from each per-syscall variant for most cases I think.

Reply:
For those cases that need to call `__wasilibc_find_relpath` twice they could just be aware of the issue and call `strdup()` before calling it a second time.   Since we control all the call sites I think its not so bad to have a few special cases like this.